Handle lookup item & attachment creation in IdentifierLookupController

mvasilak opened the pull request:
closes https://github.com/zotero/zotero-ios/issues/683

Moves item & attachment creation from identifier lookup to a separate user controller, that is guarranted to fulfill ongoing lookups, if lookup window is closed.

With this change, perhaps we should label the button as `Close`, instead of `Cancel`. Furthermore, the items added for this identifier, perhaps should indicate that they might have pending downloads.

-- 

TODO:
- [x] Keep track of separate lookup window initializations, if selected library & collections are different.
- [x] Include downloaded attachments from translated identifiers, in items toolbar label `Downloaded x/y`.
- [x] Show progress of identifier lookup in items toolbar label as `Saved x/y`.
- [x] Add button in items toolbar, that will restore lookup window, when there are identifier lookups in the background.
- [x] Add `Cancel all` button in lookup window, that will cancel all downloads and trash items/attachments already fetched.
- [x] Rename current `Cancel` button in lookup window to `Close`, that dismisses window and continues lookup in the background.
- [ ] Add a cancel button in each lookup item row, for individual cancellation. Cancelling an item, will also remove its attachments. Use an X or trash icon.
- [ ] Add a cancel button in each attachment row, that cancels/removes this attachment.
- [x] Cleanup of finished lookup items from controller, when lookup window is closed, and all operations are finished/cancelled.
- [x] When new identifiers are added for lookup, with previous ones queued, show a list of all items.

mvasilak commented on Zotero/Scenes/Detail/Lookup/LookupCoordinator.swift:
```diff
@@ -50,8 +50,12 @@ final class LookupCoordinator: NSObject, Coordinator {
-        let handler = LookupActionHandler(dbStorage: userControllers.dbStorage, fileStorage: self.controllers.fileStorage, translatorsController: self.controllers.translatorsAndStylesController,
-                                          schemaController: self.controllers.schemaController, dateParser: self.controllers.dateParser, remoteFileDownloader: userControllers.remoteFileDownloader)
+        let handler = LookupActionHandler(
+            identifierLookupController: controllers.userControllers!.identifierLookupController,
```
`userControllers` are actually passed as a non-optional parameter, somehow I missed that 😅
Check https://github.com/zotero/zotero-ios/pull/707/commits/eaeb31a0dac23c575ed08c779687364e6ac1219d

michalrentka commented:
Code-wise it's fine, it moved existing functionality to background so that things can finish even if the UI is closed. 

> With this change, perhaps we should label the button as Close, instead of Cancel

@mvasilak I think that `Cancel` currently still cancels the translation? The web view is still inside the popup, so if the popup closes, translation is cancelled, right? I'd keep it as it is - you can cancel the translation if something goes wrong, but if items are translated successfully, you can close the popup so that you don't have to watch items download one by one.

> Furthermore, the items added for this identifier, perhaps should indicate that they might have pending downloads.

@dstillman a little explanation and question: currently translation happens, item(s) are created for the main (parent) items, then files are downloaded and if individual downloads succeed then their attachment items are created. Which means if you're on slow internet you'll first see parent item(s) without attachments and only when those files are downloaded attachments start showing up.

As @mvasilak hinted, would it be better to create all items immediately and then start all downloads (and show their progress the same way as if downloading them from our server)? This approach would be more complicated just because you'd have to handle interruptions and retry downloads (if users just close the app, downloads are interrupted and will have to be restarted after the app starts again). Also if people don't really download that many big papers at once, you most likely won't even notice the difference, because smaller papers are downloaded quickly and they just show up after few seconds. So I'm not sure whether it's worth it.

dstillman commented:
Is the question about showing pending downloads in the lookup progress dialog or in the items list itself?

For the former, we know immediately from the translator whether there's an attachment, and we know the name to use. That's how we show the attachment line right away in the Zotero Connector save popup (and maybe the share extension?). So at least for that, we can show attachment progress without creating the attachment item first. Maybe we're already doing that?

If the question is about the items list, and we're talking about switching to a Close button once items are created but before downloads are done, could we hook into the lookup controller for the items list progress indicator, and also queue an 'open' action so that it opens the file once it exists? Seems very complicated, but I don't think we want to create an attachment item first, because that seems very likely to lead to attachment items with missing files, attachment items being synced before files are available, etc.

For a simpler option, could we just add an indicator/button somewhere that would reopen the lookup dialog to show progress? We do something similar for multi-item PDF metadata retrieval in the desktop app.

michalrentka commented:
> Is the question about showing pending downloads in the lookup progress dialog or in the items list itself?

The former is already implemented, it was the latter. I agree it's probably unnecessarily overcomplicated and we can create attachments only after they download and just ignore them if they don't.

> For a simpler option, could we just add an indicator/button somewhere that would reopen the lookup dialog to show progress? We do something similar for multi-item PDF metadata retrieval in the desktop app.

That's a good idea. We could have a progress bar on bottom, same as when we're downloading multiple files (screenshot below) with a button to re-open the lookup popup. That way you'd know you're currently downloading and can view state as well.

![IMG_139D1C7CA62F-1](https://github.com/zotero/zotero-ios/assets/1022956/3ed7cdc9-fd36-47be-9461-772b9ba591e6)

It could say "Lookup x/y" with progress bar below and "Open" to the right.

dstillman commented:
> We could have a progress bar on bottom, same as when we're downloading multiple files (screenshot below) with a button to re-open the lookup popup. That way you'd know you're currently downloading and can view state as well.

👍

dstillman commented:
"Lookup" is an internal term for the feature. Can it not just say "Downloaded x/y"?

michalrentka commented:
> "Lookup" is an internal term for the feature. Can it not just say "Downloaded x/y"?

Sure, I thought we might want to distinguish attachment downloads from lookup downloads. But I guess we can even combine them if both are in progress then.

michalrentka commented:
@dstillman there's one more usability issue probably. When you add multiple identifiers, downloads are now in background, but translation happens sequentially. Translation could be moved to background too (by moving the underlying webview to parent controller). Do we want to do that? Or is it fine like this?

dstillman commented:
What would the usability issue be? That translation stops if you close the app?

michalrentka commented:
> What would the usability issue be? That translation stops if you close the app?

Currently yes. The popup says "Cancel" while translations are in progress, so when you close the popup currently running translations are stopped. Started downloads finish in background.

dstillman commented:
That seems fine.

dstillman commented:
Ah, I see. So the question is if we want to change it to process them in the background with a Close button instead of (or in addition to) Cancel? And we added something to reopen the progress dialog, right? Does that button stick around after completion so you can view results?

dstillman commented:
> I guess we could add cancel to individual translations instead of general cancel and we'd keep Close for the whole window.

Why? Because some network request might be hanging? Maybe we want that, but seems weird not to also give the option to cancel the whole process if we have any cancel options.

> `translated z/w`

"Saved z/w"

michalrentka commented:
@mvasilak got it.

Some other issues:
- When single translation/download is in progress, the download progress in toolbar doesn't show up. 
https://github.com/zotero/zotero-ios/assets/1022956/41938fbd-dd2c-4518-b700-c0fe15b36fa9

- When I start translation on bad internet (I turned on network limiter), item and attachment are translated and I can see their titles, but then download progress is stuck at 0 and I don't see "downloading" in the toolbar. I see it only after it downloads a bit, do you know what it's waiting for?  
https://github.com/zotero/zotero-ios/assets/1022956/e8effc3b-35a2-4e83-b310-f63f65f95b98

michalrentka commented:
> > I guess we could add cancel to individual translations instead of general cancel and we'd keep Close for the whole window.
> 
> Why? Because some network request might be hanging? Maybe we want that, but seems weird not to also give the option to cancel the whole process if we have any cancel options.

I thought cancel would make sense when the translation finishes and you see the resulted item is not what you wanted. But we can add cancel (all) and individual cancels or just cancel option.

mvasilak commented:
@michalrentka I'd notice that as well. Probably need an initial event w/ 0 progress, similar to what `AttachmentDownloader` does. I will investigate it.

dstillman commented:
> I thought cancel would make sense when the translation finishes and you see the resulted item is not what you wanted.

The latter could be useful, but it should probably be an (x) delete button. And I guess that'd be fine for cancelling while in progress too.

mvasilak commented:
New specs, from the discussion, please verify:
* Downloaded attachments from translated identifiers, are included in items toolbar label `Downloaded x/y`.
* New label `Saved x/y` will be added also in items toolbar, to show translation progress in the background.
* Button in items toolbar, will allow to re-open queue of translated identifiers.
* Lookup window will have `Cancel all` button, that will cancel all downloads and remove items/attachments already fetched. Probably it's good to show a confirmation alert.
* Lookup window will also have `Close` button, that just continues the process in the background.
* Each lookup item row will have a cancel button, for individual cancellation. Cancelling an item, will also remove its attachments. If item already downloaded, in a sense it becomes a delete/remove button, but could be the same, e.g. an X or trash icon.
* Each attachment row will also have a cancel button, just to cancel/remove this attachment.

A few things not commented already:
* Looking up new identifiers, could open the lookup input window, but when lookup is started, add them in the list with existing lookup items, if any.
* Lookup items should be removed at some point from the list. I guess when all their attachments have finished (successfully or otherwise), and if the lookup list window is not visible at the time.

michalrentka commented:
> @michalrentka regarding the sole download not appearing, it also happens for normal downloads of attachments, due to this guard
> 
> https://github.com/zotero/zotero-ios/blob/f9bfbf88afe8b2630d135fbe04cabb1287543782/Zotero/Scenes/Detail/Items/Models/ItemsState.swift#L35
> 
> What is the reasoning behind this? Should this edge case be maintained? If so I'll make this an option for the constructor. Or could we remove this check for all cases?

Yeah it made sense when it was used just for downloads. When you tap a single attachment which needs to download, you already see the download progress in the icon, so I ignored it. I guess just remove the `guard`.

michalrentka commented:
> @michalrentka the "backgrounding" of the identifiers lookup has been implemented, and while the PR is not complete yet, I wouldn't mind some first comments.

I like the general idea, I think it's going in the right direction. There are some mostly minor things I'd comment on a finished PR, but I don't think it's worth doing actual code review until it's finished? Or do you want to fully review the `IdentifierLookupController` for example?

> * Uses web view from protocol implemented by LibrariesViewController (currently).

This is quite clean in the end so it's fine in there.

> * Uses nested function style for those private functions it makes sense, so please also provide some feedback on this.

I can't say I hate them, it's certainly better in Xcode, but it's quite hard to read when doing code review. You can't just fold them here, so when you have something like:
```
func first(...) {
    func second(...) {
        ...
    }
    func third(...) {
        ...
    }

    actual body of first
}
```
it's easy to loose track of what's what and where the actual initial (`first` in this case) function starts. Is it possible to define them on bottom instead of on top? I think that would be easier to navigate here on github at least.

mvasilak commented:
> I like the general idea, I think it's going in the right direction. There are some mostly minor things I'd comment on a finished PR, but I don't think it's worth doing actual code review until it's finished? Or do you want to fully review the `IdentifierLookupController` for example?

If they are only minor 😅, better review in total when it's finished.

> it's easy to loose track of what's what and where the actual initial (`first` in this case) function starts. Is it possible to define them on bottom instead of on top? I think that would be easier to navigate here on github at least.

That's a good suggestion, check https://github.com/zotero/zotero-ios/pull/707/commits/2d56f4964abcb1bbaa985189165c3684ff6e9674, to see if it makes it easier to read in github.

michalrentka commented:
> > I like the general idea, I think it's going in the right direction. There are some mostly minor things I'd comment on a finished PR, but I don't think it's worth doing actual code review until it's finished? Or do you want to fully review the `IdentifierLookupController` for example?
> 
> If they are only minor 😅, better review in total when it's finished.

Yes I'd leave that for now, nothing worth a bigger rewrite :).

> > it's easy to loose track of what's what and where the actual initial (`first` in this case) function starts. Is it possible to define them on bottom instead of on top? I think that would be easier to navigate here on github at least.
> 
> That's a good suggestion, check [2d56f49](https://github.com/zotero/zotero-ios/commit/2d56f4964abcb1bbaa985189165c3684ff6e9674), to see if it makes it easier to read in github.

Yup I like this more. If you're ok with it we can stick with that.

mvasilak commented:
@dstillman @michalrentka only the individual cancellation actions are not yet implemented. As this is getting quite big and complex, it might be could to review the functionality & code so far, and even consider moving the final tasks to a separate PR. 

Some notes:
* Looking up identifiers, while there are ongoing lookups that have been dismissed before, will show a list of all lookups, with the new ones added on top.
* Cancelling all lookups will stop unfinished operations, and trash items/attachments already stored.
* Cancelling all lookups will restore the lookup window to the prior textview input.
* To restore ongoing lookups, the progress view in the items toolbar has an indicative UI and uses a tap gesture recognizer.

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+    // MARK: Setups
+    private func setupObservers() {
+        remoteFileDownloader.observable
+            .subscribe { update in
```
you can create scheduler with `backgroundQueue` and just call `observeOn(.backgroundScheduler)` instead of calling `backgroundQueue.async` down in `finish(download:attachment:)`.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/cac7e7339967ac0f968550762198127a4132c069

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+    
+    func lookUp(libraryId: LibraryIdentifier, collectionKeys: Set<String>, identifier: String) {
+        let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+        lookupWebViewHandlersByLookupSettings[lookupSettings]?.lookUp(identifier: identifier)
```
Does this always exist? Maybe add a `guard` and a `DDLogError` at least in case something breaks in future and users report that this does nothing.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/97438bb929d4e0d770d16970d92326c45fac536b

michalrentka commented on Zotero/Scenes/Master/Libraries/Views/LibrariesViewController.swift:
```diff
@@ -207,3 +212,18 @@ extension LibrariesViewController: UITableViewDataSource, UITableViewDelegate {
+    }
+    
+    func removeWebView(_ webView: WKWebView) {
+        if view.subviews.contains(where: { $0 == webView }) {
```
Do you need to check this? Does this even need to be called here? Can't you just call `webView.removeFromSuperview()` directly?

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/c79834da3fe9b9acead80374f8016ff0327d007b

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+            for key in keys {
+                if let webView = self.lookupWebViewHandlersByLookupSettings.removeValue(forKey: key)?.webViewHandler.webView {
+                    inMainThread {
+                        self.webViewProvider?.removeWebView(webView)
```
Can't we call `webView.removeFromSuperview()` directly?

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/c79834da3fe9b9acead80374f8016ff0327d007b

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+            DDLogInfo("IdentifierLookupController: cancel all lookups")
+            let keys = self.lookupWebViewHandlersByLookupSettings.keys
+            for key in keys {
+                if let webView = self.lookupWebViewHandlersByLookupSettings.removeValue(forKey: key)?.webViewHandler.webView {
```
I'd change this to `guard`, at least we'll get rid of some depth below.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/1903e2ba890cb62198179c812fb54e43b0112a4a

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+                    return nil
+                }
+            }
+            for (response, libraryId) in storedItemResponses {
```
Call `self.backgroundQueue.async` before the `for` loop, unnecessary overhead. You can also create all dbrequests in the `for` loop (or `map`) and then call single `dbStorage.perform` with all requests for single db write.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/f73cdf75e9b42a27d48d096deeff37488fd0989d

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+    }
+    
+    func getIdentifiersLookupCount(callback: @escaping (Int, Int, Int, [LookupData]) -> Void) {
+        accessQueue.async { [weak self] in
```
Does this need to be asynchronous? If yes, this looks overly complicated, I'd just do
```
accessQueue.async { [weak self] in
    guard let self else { return }
    callback(self.lookupSavedCount, self.lookupFailedCount, self.lookupTotalCount, self.lookupData)
}
```

Ideally it would be synchronous and then you'd need all those variables.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/7b9aa5de1ea91ae6227dbbbd0bd94261c2e0745d

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+    }
+    
+    // MARK: Lookup Data
+    func cleanupLookup(force: Bool, alreadyInQueue: Bool, update: Update?) {
```
I don't love the `alreadyInQueue` parameter. You specify `_cleanupLookup` function anyway, so I'd just have separate functions, where `cleanupLookup` calls `_cleanupLookup` on proper queue and if you know you're on proper queue you can call `_cleanupLookup` directly.

I'd also change the name to `cleanupLookupIfNeeded` since you check for finished tasks in it.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/1196ff95e7aa78d9f47d4d5b321dd5dd96f0c6a4

michalrentka replied:
Didn't know you can check queue like that, nice :)

michalrentka commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+        }
+        
+        func _cleanupLookup(force: Bool, update: Update?) {
+            var cleanup = false
+            if force {
+                cleanup = true
+            } else if lookupRemainingCount == 0, remoteFileDownloader.batchData.2 == 0 {
+                if let presenter {
+                    inMainThread(sync: true) {
+                        cleanup = !presenter.isPresenting()
+                    }
+                } else {
+                    cleanup = true
+                }
+            }
+            guard cleanup else { return }
+            // If forced,
+            // or all lookups have been saved or failed, all attachments have finished, and either no presenter is assigned, or it doesn't present currently,
+            // then cleanup.
+            lookupData = []
+            lookupSavedCount = 0
+            lookupFailedCount = 0
+            DDLogInfo("IdentifierLookupController: cleaned up lookup data")
+            guard let update else { return }
+            observable.on(.next(update))
+        }
```
```suggestion
            if force {
                cleanup(update: update)
            }
            if lookupRemainingCount == 0 && remoteFileDownloader.batchData.2 == 0 {
                // If all lookups have been saved or failed, all attachments have finished
                guard let presenter else {
                    // and no presenter is assigned, then cleanup
                    cleanup(update: update)
                    return
                }
                // and presenter is assigned
                var isPresenting = false
                inMainThread(sync: true) {
                    isPresenting = presenter.isPresenting()
                }
                if !isPresenting {
                    // but is not presenting, then cleanup
                    cleanup(update: update)
                }
            }
            
            func cleanup(update: Update?) {
                lookupData = []
                lookupSavedCount = 0
                lookupFailedCount = 0
                DDLogInfo("IdentifierLookupController: cleaned up lookup data")
                guard let update else { return }
                observable.on(.next(update))
            }
        }
```

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/46ae8d700e03a23b5cdc03ac6026b9f5dee80234

mvasilak commented on Zotero/Controllers/Attachment Downloader/RemoteAttachmentDownloader.swift:
```diff
@@ -80,38 +99,66 @@ final class RemoteAttachmentDownloader {
     func download(data: [(Attachment, URL, String)]) {
-        self.queue.async(flags: .barrier) { [weak self] in
-            guard let self = self else { return }
+        self.accessQueue.async(flags: .barrier) { [weak self] in
```
`createDownload` that is called within the block modifies local properties. Moved it as a nested function to make it more clear. Check https://github.com/zotero/zotero-ios/pull/707/commits/018260333c910c4086b3674f8b53b23d041e1796

mvasilak commented on Zotero/Scenes/Detail/Lookup/ViewModels/LookupActionHandler.swift:
```diff
@@ -11,90 +11,94 @@ import Foundation
+                                state.lookupState = .failed(error)
+                            }
+
+                        case .noIdentifiersDetected:
```
`.lookupError` is used for errors observed from `LookupWebViewHandler`, while `.noIdentifiersDetected` is not considered an error in the `IdentifierLookupController` context. I modified it in `LookupActionHandler`, because it is handled as an error, to display the failure message. Do you think we should merge them both in one case?

michalrentka replied:
I'd maybe consider no identifiers detected as error, because the input was most likely wrong then. But if there are valid reasons why it's not detected while not being an error this is fine then.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/a29a864c86ff0eabc79981d05550e39492fdba52

It merges the `identifiersDetected` and `noIdentifiersDetected` states, and lets `LookupActionHandler` to determine how the view model should be affected when the identifiers array passed with the update is empty.

mvasilak commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+            inMainThread(sync: true) {
+                if let webView = self.webViewProvider?.addWebView() {
+                    let lookupWebViewHandler = LookupWebViewHandler(lookupSettings: lookupSettings, webView: webView, translatorsController: self.translatorsController)
+                    self.lookupWebViewHandlersByLookupSettings[lookupSettings] = lookupWebViewHandler
```
`WKWebView` construction needs to happen in main thread, otherwise it crashes.

michalrentka replied:
Sure, but `lookupWebViewHandlersByLookupSettings` should be accessed/written from `accessQueue`. I guess in this case it doesn't matter since it's synchronous and inside `accessQueue` barrier block. I'd probably still keep web view initialization on main queue and assign it on `accessQueue`, but this shouldn't create any race conditions.

mvasilak replied:
Check https://github.com/zotero/zotero-ios/pull/707/commits/60b3c43d1a29b3e89e2967aad066e81f71b02bb0

mvasilak commented on Zotero/Controllers/Attachment Downloader/RemoteAttachmentDownloader.swift:
```diff
@@ -80,38 +99,66 @@ final class RemoteAttachmentDownloader {
-                guard let self = self else { return }
+            self?.processingQueue.async(flags: .barrier) {
+                guard let self else { return }
                 self.finish(download: download, file: file, attachment: attachment, parentKey: parentKey, result: result)
```
Correct, check https://github.com/zotero/zotero-ios/pull/707/commits/cb8af619324e87b09b192b0a566960bdc2f325f7

mvasilak commented on Zotero/Scenes/Detail/Lookup/ViewModels/LookupActionHandler.swift:
```diff
@@ -11,90 +11,94 @@ import Foundation
-                        self?.finish(download: update.download, attachment: attachment, in: viewModel)
-                    case .cancelled, .failed, .progress: break
+        case .initialize:
+            identifierLookupController.initialize(libraryId: libraryId, collectionKeys: collectionKeys) { [weak self] lookupData in
```
Check https://github.com/zotero/zotero-ios/pull/707/commits/87174e01219837a21b0e7d7d9239c66dd0308be7

mvasilak commented on Zotero/Controllers/IdentifierLookupController.swift:
```diff
@@ -0,0 +1,445 @@
+                    }
+                    
+                case .item(let data):
+                    guard let lookupId = data["identifier"] as? [String: String] else { return }
```
I made it a warning, should it be an error instead?
Check https://github.com/zotero/zotero-ios/pull/707/commits/83db6f69d8dd797de76a743c13c55c03f9b9c64d

michalrentka replied:
Warning is fine

mvasilak commented on Zotero/Scenes/Detail/Lookup/ViewModels/LookupActionHandler.swift:
```diff
@@ -106,82 +110,22 @@ final class LookupActionHandler: ViewModelActionHandler, BackgroundDbProcessingA
-                        let failedData = LookupState.LookupData(identifier: identifier, state: .failed)
-                        self.update(lookupData: failedData, in: viewModel)
-                    }
+                    state.lookupState = .lookup(data)
```
Check https://github.com/zotero/zotero-ios/pull/707/commits/a29a864c86ff0eabc79981d05550e39492fdba52

To make things hopefullly simpler:
* A separate `LookupState` for `waitingInput` was added, and `loadingIdentifiers` is used as before just for waiting until the first updates arrive. 
* Since `IdentifierLookupController` needs to hold the lookup state anyway, it passes it to every update it publishes, so `LookupActionHandler` just needs to copy it when needed.
* Also fixed some potential edge case bugs in IdentifierLookupController, by making the workflow for publishing updates more strict.

michalrentka commented:
Code looks good to me now, there are a few issues I noticed when testing:

1. Looks a bit weird that one part is bigger than the other, I'd probably make both small:
![IMG_0408](https://github.com/zotero/zotero-ios/assets/1022956/54ab566f-a656-4ca1-b009-6c307da1ecf6)

2. When I add multiple identifiers, the popup correctly scales and takes maximum possible height. But when I close the popup and re-open with multiple items, it doesn't scale up properly:
![IMG_0409](https://github.com/zotero/zotero-ios/assets/1022956/f15cc422-364e-4039-9d2e-7185dabd9ee3)

3. When downloads are in progress, the table view reloads for progress updates are creating this weird blinking:
https://github.com/zotero/zotero-ios/assets/1022956/3d88f21f-54d9-4ee3-a8cf-dbc85412c01f

mvasilak commented:
@michalrentka 

>     1. Looks a bit weird that one part is bigger than the other, I'd probably make both small:

I differentiated them in purpose, to make one more button like :) 
Probably the color change suffices, I'll make them the same size.

>     2. When I add multiple identifiers, the popup correctly scales and takes maximum possible height. But when I close the popup and re-open with multiple items, it doesn't scale up properly:

I will investigate this one.

>     3. When downloads are in progress, the table view reloads for progress updates are creating this weird blinking:
>        https://github.com/zotero/zotero-ios/assets/1022956/3d88f21f-54d9-4ee3-a8cf-dbc85412c01f

If I'm not mistaken, that is a prexisting problem. I suggest we open a separate issue.

michalrentka commented:
> > ```
> > 3. When downloads are in progress, the table view reloads for progress updates are creating this weird blinking:
> >    https://github.com/zotero/zotero-ios/assets/1022956/3d88f21f-54d9-4ee3-a8cf-dbc85412c01f
> > ```
> 
> If I'm not mistaken, that is a prexisting problem. I suggest we open a separate issue.

That's possible :) feel free to open new issue

mvasilak commented:
@michalrentka ready for another look, regarding the UI issues you noticed.

michalrentka commented:
@mvasilak I've encountered some more weirdness, but it works better now.

1. First I accidentally pasted the same 4 identifiers twice and translation got stuck at some point:
![IMG_0411](https://github.com/zotero/zotero-ios/assets/1022956/b69395de-bcbf-4523-ab8c-2ec78cc05bc3)

2. Then I tried again with original 4 identifiers and the first one never translated as well. If there is an error it should show it:
![IMG_0412](https://github.com/zotero/zotero-ios/assets/1022956/2ad7a0a4-0bc8-4e0b-9ab2-ccc0082f791a)

3. And then when I tried `Cancel All` button I was shown the initial popup with input, so I tried adding the same identifiers and starting translation, but then it just got stuck at the beginning:
![IMG_0413](https://github.com/zotero/zotero-ios/assets/1022956/eceeec7e-621c-4ea2-a575-96f79a378f6b)

I was testing with these identifiers: "arXiv:2307.06949, arXiv:2307.06925, arXiv:2307.06526, arXiv:2307.06335"

mvasilak commented:
@michalrentka latest issues have been addressed and PR is ready for another review. Changes of note:
* When canceling all the lookup window is dismissed. If we want to keep it open, some minor changes are needed, to re-initialize the web views, as they are now more aggresively cleaned up when all identifiers have been processed.
* Error messages have been improved to accommodate the case where there are ongoing lookups in the background.

mvasilak commented:
Rebased to `master`

mvasilak commented:
Rebased to `master`. @michalrentka @dstillman I wouldn't mind another quick look (nothing changed, just in case you find an issue due to other changes since last time). I think this is ready for release, and I propose to move the remaining taks to another issue, with low priority, as they are just covering edge use cases.

michalrentka reviewed (approved):
I think it's good to merge this, it definitely improves and fixes some things and we released necessary bug fixes that we needed out quickly. @dstillman any thoughts? Otherwise I'll merge this next week.
